packages/@lwc/template-compiler/src/parser/expression-complex/html.ts

```diff
@@ -0,0 +1,267 @@
+
+            const expressionTextNodeValue = this.parseTemplateExpression();
+
+            // Create a new text-node token to contain our `{expression}`
+            this._createCharacterToken(Tokenizer.CHARACTER_TOKEN, expressionTextNodeValue);
+
+            // Emit the text node token containing the `{expression}`
+            this._emitCurrentCharacterToken();
+
+            // Moving the cursor back by one allows the state machine to correctly detect
+            // the state into which it should next transition.
+            this._unconsume();
+        } else {
+            super.DATA_STATE(codePoint);
+        }
+    }
+}
+
+interface TemplateHtmlParserOptions extends ParserOptions {
+    preparsedJsExpressions: PreparsedExpressionMap;
+}
+
+/**
+ * This class extends `parse5`'s internal parser. The heavy lifting is
+ * done in the tokenizer. This class is only present to facilitate use
+ * of that tokenizer when parsing expressions.
+ */
+class TemplateHtmlParser extends Parser {
+    preparsedJsExpressions: PreparsedExpressionMap;
+    tokenizer!: TemplateHtmlTokenizer;
+
+    constructor(_options: TemplateHtmlParserOptions) {
+        const { preparsedJsExpressions, ...options } = _options;
+        super(options);
+        this.preparsedJsExpressions = preparsedJsExpressions;
+    }
+
+    _bootstrap(document: Document, fragmentContext: any) {
+        super._bootstrap(document, fragmentContext);
+        // The default `_bootstrap` method creates a new Tokenizer; here, we ensure that our
+        // customized tokenizer is used.
+        this.tokenizer = new TemplateHtmlTokenizer(this);
+    }
+
+    // The parser will try to concatenate adjacent text tokens into a single
+    // text node. Template expressions should be encapsulated in their own
+    // text node, and not combined with adjacent text or whitespace. To avoid
+    // that, we create a new text node for the template expression rather than
+    // allowing the concatenation to proceed.
+    _insertCharacters(token: any) {
+        if (token.chars[0] !== '{') {
+            return super._insertCharacters(token);
+        }
+        const parentNode = this.openElements.current;
+        const textNode: TextNode = {
+            nodeName: '#text',
+            value: token.chars,
+            sourceCodeLocation: token.location,
+            parentNode,
+        };
+        parentNode.childNodes.push(textNode);
+    }
+}
+
+interface ParseFragmentConfig {
+    ctx: ParserCtx;
+    sourceCodeLocationInfo: boolean;
+    onParseError: (error: ParsingError) => void;
+}
+
+/**
+ * Parse the LWC template using a customized parser & lexer that allow
+ * for template expressions to be parsed correctly.
+ *
+ * @param      {string}               source  raw template markup
+ * @param      {ParseFragmentConfig}  config
+ *
+ * @return     {DocumentFragment}     the parsed document
+ */
+export function parseFragment(source: string, config: ParseFragmentConfig): DocumentFragment {
```

Comment:
If `ctx` was passed in as a third param, creating `interface ParseFragmentConfig` can be avoided.